python/pyarrow/_flight.pyx

```diff
@@ -413,18 +407,63 @@ cdef class FlightDataStream:
 
 
 cdef class RecordBatchStream(FlightDataStream):
-    def __init__(self, reader):
-        # TODO: we don't really expose the readers in Python.
-        pass
+    """A Flight data stream backed by RecordBatches."""
+    def __init__(self, data_source):
+        """Create a RecordBatchStream from a data source.
 
-
-cdef void _get_flight_info(void* self, CFlightDescriptor c_descriptor,
-                           unique_ptr[CFlightInfo]* info):
-    """Callback for implementing Flight servers in Python."""
-    raise NotImplementedError("GetFlightInfo is not implemented")
+        Parameters
+        ----------
+        data_source : RecordBatchReader or Table
+        """
+        cdef shared_ptr[CRecordBatchReader] reader
+        if isinstance(data_source, _CRecordBatchReader):
+            reader = (<_CRecordBatchReader> data_source).reader
+            stream = new CRecordBatchStream(reader)
+            self.stream.reset(stream)
+        elif isinstance(data_source, lib.Table):
+            table = (<Table> data_source).table
+            reader.reset(new TableBatchReader(deref(table)))
+            stream = new CRecordBatchStream(reader)
+            self.stream.reset(stream)
+        else:
+            raise TypeError("Expected RecordBatchReader or Table, "
+                            "but got: {}".format(type(data_source)))
 
 
-cdef void _do_put(void* self, unique_ptr[CFlightMessageReader] reader):
+cdef int _list_flights(void* self, const CCriteria* c_criteria,
+                       unique_ptr[CFlightListing]* listing) except -1:
+    """Callback for implementing ListFlights in Python."""
+    cdef:
+        vector[CFlightInfo] flights
+    result = (<object> self).list_flights(c_criteria.expression)
+    for info in result:
+        if not isinstance(info, FlightInfo):
+            raise TypeError("FlightServerBase.list_flights must return "
+                            "FlightInfo instances, but got {}".format(
+                                type(info)))
+        flights.push_back(CFlightInfo(deref((<FlightInfo> info).info.get())))
+    listing.reset(new CSimpleFlightListing(flights))
+    return 0
+
+
+cdef int _get_flight_info(void* self, CFlightDescriptor c_descriptor,
+                          unique_ptr[CFlightInfo]* info) except -1:
+    """Callback for implementing Flight servers in Python."""
+    cdef:
+        FlightDescriptor py_descriptor = \
+            FlightDescriptor.__new__(FlightDescriptor)
+    py_descriptor.descriptor = c_descriptor
+    result = (<object> self).get_flight_info(py_descriptor)
+    if not isinstance(result, FlightInfo):
+        raise TypeError("FlightServerBase.get_flight_info must return "
+                        "a FlightInfo instance, but got {}".format(
+                            type(result)))
+    info[0] = move((<FlightInfo> result).info)
```

Comment:
Using `std::move` may be error-prone, if the FlightServer implementation keeps persistent FlightInfo objects around. Our preferred idiom accross Arrow is to use `shared_ptr` to prevent this kind of issues. `unique_ptr` can avoid performance issues, but that's hardly ever a concern (just don't create / destroy `shared_ptr`s in a critical loop).

Reply:
I'll change these to copy the object, then. Sharing an object via `shared_ptr` won't work as the Flight API requires a `unique_ptr`.